enhancements/kubelet/user-namespaces-support.md

```diff
@@ -0,0 +1,339 @@
+
+### Goals
+
+- Enable Openshift users to request a pod be put in a user namespace
+- Update SCC to take user namespaces into account when choosing the security profile of a container
+- Add support for users to run podman within an Openshift pod without being privileged.
+
+### Non-Goals
+
+- Anything not related to user namespaces or running nested containers.
+
+## Proposal
+
+There are two pieces to this proposal:
+- Extend SCC to be aware of the `hostUsers` field:
+    - Add a new field `ContainerUserPolicy` to SCC, which will be `AllowHostNamespaces` by default
+    - Add a new SCC to the default list: `restricted-v3`
+        - This SCC will be identical to `restricted-v2`, but have `ContainerUserPolicy` set to `RequirePodNamespace`
+    - Add a new SCC to the default list: `nonroot-v3`
+        - This SCC will be identical to `nonroot-v2`, but have `ContainerUserPolicy` set to `RequirePodNamespace`
+    - Add a new SCC to the default list: nested-container. It will have:
+        - SELinux context set to `MustRunAs.Type: container_engine_t`
+        - `RunAsUserStrategy: RunAsAny`
+        - `UserNamespacesPolicy`: `RequirePodNamespace`
+        - And otherwise mirror the `restricted-v2` profile
+- Add the features `UserNamespacesSupport`, `UserNamespacesPodSecurityStandards` and `ProcMountType` to the list that qualify a cluster as `TechPreviewNoUpgrade`
+    - While `UserNamespacesSupport` is beta in upstream Kubernetes 1.30, it is off by default.
+
+### Workflow Description
+
+- (TechPreview) A cluster admin turns on the `UserNamespacesSupport`, `UserNamespacesPodSecurityStandards` and `ProcMountType` features with the featuregate CRD.
+- The cluster reconciles, enabling these features in the kubelet and apiservers.
+- Pod authors can update their pods and set the `hostUsers` field to `false`.
+    - Depending on the SCC the namespace is allowed to use, the pod author may be required to do so.
+- Pod authors can also set the `procMount` field to `Unmasked` (if they also set `hostUsers` to `false`).
+- Pod authors could then run nested containers, or other operations that can benefit from a user namespace and unmasked `/proc`
+
+### API Extensions
+
+- Add `UserNamespacesPolicy` to SCC. This relies on approval from the apiserver team.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+N/A
+
+#### Standalone Clusters
+
+N/A
+
+#### Single-node Deployments or MicroShift
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+#### SecurityContextConstraint Updates
+
+##### ``UserNamespacesPolicy`` field
+
+Add the following to the [Openshift API](https://github.com/openshift/api/blob/76a71dac36a08eab1b240c6c8d4e39c813b1b12b/security/v1/types.go):
+
+```
+       // userNamespacesPolicy determines if the policy allows host users in containers.
+       // Valid values are "AllowHostNamespaces", "RequirePodNamespaces", and omitted.
```

Comment:
It would be good to explain what AllowHostNamespaces and RequirePodNamespaces mean from a pod spec perspective. One of these means I have to set hostUsers false or something right? The other means it can be true or false?